examples/bpi-monolith-test-suite/bpi/components/zkcircuits/circuitlib/orderCircuit.go

```diff
@@ -0,0 +1,183 @@
+	OrderSalt          frontend.Variable `gnark:",secret"`
+	BuyerPK            frontend.Variable `gnark:",secret"`
+	AgreementStateRoot frontend.Variable `gnark:",secret"`
+	AgreementStateSalt frontend.Variable `gnark:",secret"`
+}
+
+type Order struct {
+	ProductId frontend.Variable `gnark:",secret"`
+	BuyerSig  frontend.Variable `gnark:",secret"`
+}
+
+var verifyingKey string
+
+func (circuit *OrderCircuit) Define(api frontend.API) error {
+	mimc, _ := mimc.NewMiMC(api)
+
+	// Asserts Input AgreementStateCommitment and Derived AgreementStateCommitment are equal
+	sum := api.Add(circuit.PI.AgreementStateRoot, circuit.PI.AgreementStateSalt)
+	mimc.Write(sum)
+	api.AssertIsEqual(circuit.I.AgreementStateCommitment, mimc.Sum())
+
+	// Asserts Input StateObjectCommitment and Derived StateObjectCommitment are equal
+	mimc.Reset()
+	sum = api.Add(circuit.PI.OrderRoot, circuit.PI.OrderSalt)
+	mimc.Write(sum)
+	api.AssertIsEqual(circuit.I.StateObjectCommitment, api.Add(circuit.PI.OrderRoot, circuit.PI.OrderSalt))
+
+	//Asserts CalculatedAgreementRoot and PrivateInput AgreementStateRoot are equal
+	api.AssertIsEqual(circuit.I.CalculatedAgreementRoot, circuit.PI.AgreementStateRoot)
+
+	//Verifies Signature is correct - mimics mock fashion
+	api.AssertIsEqual(circuit.PI.O.BuyerSig, api.Add(circuit.PI.BuyerPK, circuit.PI.O.ProductId))
+
+	return nil
+}
+
+func GenerateProof(p json.Input) (groth16.Proof, string) {
+	var zkcircuit OrderCircuit
+
+	r1cs, err := frontend.Compile(ecc.BN254, backend.GROTH16, &zkcircuit)
+	if err != nil {
+		fmt.Println(err)
+		return nil, "Invalid circuit compilation"
+	}
+
+	assignment := assign(p)
+	witness, err := frontend.NewWitness(assignment, ecc.BN254)
+	if err != nil {
+		fmt.Println(err)
+		return nil, "Witness Creation Failed!"
+	}
+
+	pk, vk, err := groth16.Setup(r1cs)
+	if err != nil {
+		fmt.Println(err)
+		return nil, "Groth16 Setup Failed!"
+	}
+
+	buf := new(bytes.Buffer)
+	_, err = vk.(io.WriterTo).WriteTo(buf)
+	verifyingKey = hex.EncodeToString(buf.Bytes())
+
+	proof, err := groth16.Prove(r1cs, pk, witness)
+	if err != nil {
+		fmt.Println(err)
+		return nil, "Proof Creation Failed!"
+	} else {
+		return proof, ""
+	}
+}
+
+func VerifyProof(v json.Input, proof interface{}) string {
+	assignment := assign(v)
+
+	witness, _ := frontend.NewWitness(assignment, ecc.BN254)
+	publicWitness, err := witness.Public()
+	if err != nil {
+		fmt.Println(err)
+		return "Witness Creation Failed!"
+	}
+
+	vk, err := hex.DecodeString(verifyingKey)
+	if err != nil {
+		fmt.Println(err)
+		return "Verifying Key Decoding Failed!"
+	}
+
+	_vk, err := DeserializeVerifyingKey(vk)
+	if err != nil {
+		fmt.Println(err)
+		return "Verifying Key Deserialization Failed!"
+	}
+
+	err = groth16.Verify(proof.(groth16.Proof), _vk.(groth16.VerifyingKey), publicWitness)
+	if err != nil {
+		fmt.Println(err)
+		return "Invalid verification!"
+	} else {
+		return "Valid Proof"
+	}
+}
+
+func assign(p json.Input) *OrderCircuit {
```

Comment:
we can rename to NewOrderCircuit instead of assign, assign is too generic

Reply:
renamed to assignCircuitInputs